### DISCUSSION TITLE
Is Buffer class thread-safe?

### DISCUSSION THREAD

**Discussion by glcrazier** [apache/skywalking#7633] [Q&A] [answered]
When I read source code of skywalking agent core, I notice that the Buffer class doesn't have any synchronization on object array read and write operations. Would it be a problem if the buffer is heavily written by different threads? Would it be better to use ArrayBlockingQueue to implement  the same functionality without blocking access?

source code link: org/apache/skywalking/apm/commons/datacarrier/buffer/Buffer.java

**Comment by wu-sheng** [comment, marked as answer]
I don't know what do you mean not thread safe. They are safe.

**Reply by glcrazier** [reply]
If write thread switches inside save method and read thread clears buffers, it has possibilities that the buffer is empty but save method returns false which cause to discard trace segments. Is it expected?

**Reply by wu-sheng** [reply]
Why buffer's slot can't be identified correctly? What causes it?

**Reply by glcrazier** [reply]
Below is save method:
```java
    public boolean save(T data) {
        int i = index.getAndIncrement();
        if (buffer[i] != null) {
// if thread switches here
            switch (strategy) {
                case IF_POSSIBLE:
                    return false;
                default:
            }
        }
        buffer[i] = data;
        return true;
    }
```
obtain method:
```java
    void obtain(List<T> consumeList, int start, int end) {
        for (int i = start; i < end; i++) {
            if (buffer[i] != null) {
                consumeList.add((T) buffer[i]);
                buffer[i] = null;
            }
        }
    }
```
If thread switches after identifying a slot is occupied in save method and obtain method resets all slots, then save method continues executing as the data is dropped which in fact the slot is available at this moment. Is my understanding correct?

**Reply by wu-sheng** [reply]
Yes, and that is expected. The throughput has higher priority, we don't want to lock. It would be a disaster to agent.

**Reply by glcrazier** [reply]
Thanks for your reply~
